cy.selectFile docs (#4241)
cypress-io/cypress-documentation: content/guides/references/migration-guide.md

title: Migration Guide
---

## Migrating to Cypress 10.0

This guide details the changes and how to change your code to migrate to Cypress
10.0. [See the full changelog for 8.0](/guides/references/changelog#10-0-0).

## Attaching Files

Attaching files to input elements or dropping them over the page is available

> in Cypress 10.0. Read the [`cy.attachFile()` API docs](/api/commands/attachFile)

Reviewer: Camelcase will 404 here

api/commands/attachfile
```suggestion
in Cypress 10.0. Read the [`cy.attachFile()` API docs](/api/commands/attachFile)
```
```suggestion
in Cypress 10.0. Read the [`cy.attachFile()` API docs](/api/commands/attachFile)
```

Author: Fixing.

Note that urls are actually case-insensitive - https://DOCS.cyPrEss.iO/api/cOmMaNdS/wItHiN works just fine. :D

Reviewer: Ah ok. Likely because I was looking at this before the addition to the sidebar.json and assumed that was the issue. I'm going to write all my URLs in this fashion moving forward 😆

for more information on how this works and how to use it.

> The
> [`cypress-file-upload`](https://github.com/abramenal/cypress-file-upload)
> plugin has been deprecated in favor of `attachFile()` built into Cypress.
> There's guidance below on how to migrate from the
> [`cypress-file-upload`](https://github.com/abramenal/cypress-file-upload)
> plugin to Cypress's built-in attachFile command.

Reviewer 2: ```suggestion
The
[`cypress-file-upload`](https://github.com/abramenal/cypress-file-upload)
plugin has been deprecated in favor of `.attachFile()` built into Cypress.
There's guidance below on how to migrate from the
[`cypress-file-upload`](https://github.com/abramenal/cypress-file-upload)
plugin to Cypress's built-in attachFile command.
```

> #### Quick guide
> In the first argument:

Reviewer 2: ```suggestion
#### Quick guide

The argument signature is different for `.attachFile()` than the `cypress-file-upload` plugin. You can follow the steps below for each argument in order to migrate:

In the first argument:
```

- `filePath`: Prefix value with `fixtures/'. Rename to `contents`.
- `fileContent`: Use `Buffer.from()` rather than `Cypress.Blob` methods. Rename to `contents`.
- `encoding`: Remove. No longer needed due to improved binary file handling in Cypress 9.0.
- `mimeType`: Remove. No longer needed due to improved binary file handling in Cypress 9.0.

In the second argument:
- `subjectType`: Rename to `action`.

> - `allowEmpty`: Remove. `cy.attachFile()` does not check the length of a file read from disk, only its existence.

Reviewer: ```suggestion
- `allowEmpty`: Remove. `.attachFile()` does not check the length of a file read from disk, only its existence.
```

`cypress-file-upload`

```js
cy.get('[data-cy="dropzone"]').attachFile({
  filePath: 'myfixture.json',
  fileName: 'customFileName.json',
})
```

<Badge type="success">After</Badge> Attaching a fixture from disk with
Cypress 10.0. Cypress follows paths from the root of your test folder
(as [`cy.readFile()`](/api/commands/readfile)).

```js
cy.get('[data-cy="dropzone"]').attachFile({
  contents: 'fixtures/myfixture.json',
  fileName: 'customFileName.json',
});
```

#### Working with file contents

<Badge type="danger">Before</Badge> Working with file contents before attaching using
`cypress-file-upload`

```js
const special = 'file.spss';

cy.fixture(special, 'binary')
  .then(Cypress.Blob.binaryStringToBlob)
  .then(fileContent => {
    // ...process file contents
    cy.get('[data-cy="file-input"]').attachFile({
      fileContent,
      filePath: special,
      encoding: 'utf-8',
      lastModified: new Date().getTime(),
    });
  });
```

<Badge type="success">After</Badge> Working with file contents before attaching with
Cypress 10.0. The `null` encoding introduced in Cypress 9.0 makes
working with binary data simpler, and is the preferred encoding for use with

> `cy.attachFile()`.

Reviewer: ```suggestion
`.attachFile()`.
```

CONTENT_TBD

### Attaching Files

Attaching files to input elements or dropping them over the page is available
in Cypress 10.0. Read the [`.attachFile()` API docs](/api/commands/attachfile)
for more information on how this works and how to use it.

The
[`cypress-file-upload`](https://github.com/abramenal/cypress-file-upload)
plugin has been deprecated in favor of `.attachFile()` built into Cypress.
There's guidance below on how to migrate from the
[`cypress-file-upload`](https://github.com/abramenal/cypress-file-upload)
plugin to Cypress's built-in attachFile command.

#### Quick guide

> The argument signature is different for `.attachFile()` than the `cypress-file-upload` plugin. You can follow the steps below for each argument in order to migrate:

Reviewer: This is technically correct. I've only learned this recently because I have kids but they hate when people correct them on it 🤣 

```suggestion
The argument signature is different for `.attachFile()` from the `cypress-file-upload` plugin. You can follow the steps below for each argument in order to migrate:
```

Author: It's not different for attachFiles() *from* cypress-file-upload though, that would mean something different!

I'll tweak the working a bit.

In the first argument:

> - `filePath`: Prefix value with `fixtures/'. Rename to `contents`.

Reviewer: Was formatting awkwardly. I believe the second line needed to be removed, right? In favor of the next line? 

```suggestion
- `filePath`: Prefix value with `fixtures/`.
```

Author: I added the missing tick.

In cypress-flie-upload, you could only use one at a time - providing both together was an error. So there are really two cases: Rename filePath to contents, *or* rename fileContent to contents. I'm open to better ways to explain that in the migration guide, but I think it would be clear enough when a user sees their code

```
{
  filePath: 'foo.json',
  encoding: 'utf-8',
}
```

that it turns into `{contents: 'fixture/foo.json'}` and that when they have

```
{
  fileContent: Blob.atob([91, 10]),
  mimeType: 'text/plain',
}
```

turns into `{content: Buffer.from([91, 10])}`.

I can't provide specific guidance on the Blob -> Buffer change because it's very application specific.

Reviewer: Got it. thx!

- `fileContent`: Use `Buffer.from()` rather than `Cypress.Blob` methods. Rename to `contents`.
- `encoding`: Remove. No longer needed due to improved binary file handling in Cypress 9.0.
- `mimeType`: Remove. No longer needed due to improved binary file handling in Cypress 9.0.

In the second argument:
- `subjectType`: Rename to `action`.
- `allowEmpty`: Remove. `.attachFile()` does not check the length of a file read from disk, only its existence.

### Examples

Below are several examples of migrating various commands from `cypress-file-upload`
to the builtin `.attachFile()`.

#### Read and attach a fixture

<Badge type="danger">Before</Badge> Attaching a fixture from disk with
`cypress-file-upload`

```js
cy.get('[data-cy="file-input"]').attachFile('myfixture.json');
```

<Badge type="success">After</Badge> Attaching a fixture from disk with
Cypress 10.0. Cypress follows paths from the root of your test folder

> (as [`cy.readFile()`](/api/commands/readfile)).

Reviewer: Might be more clear (if only slightly)

```suggestion
(same as [`cy.readFile()`](/api/commands/readfile)).
```

```js
cy.get('[data-cy="file-input"]').attachFile('fixtures/myfixture.json');

// Or

cy.fixture('myfixture.json', { encoding: null }).as('myfixture')
cy.get('[data-cy="file-input"]').attachFile('@myfixture');

```

#### Using drag-n-drop

<Badge type="danger">Before</Badge> Dragging and dropping a file with
`cypress-file-upload`

```js
cy.get('[data-cy="dropzone"]').attachFile('myfixture.json', { subjectType: 'drag-n-drop' })
```

<Badge type="success">After</Badge> Attaching a fixture from disk with
Cypress 10.0. Cypress follows paths from the root of your test folder

> (as [`cy.readFile()`](/api/commands/readfile)).

Reviewer: ```suggestion
(same as [`cy.readFile()`](/api/commands/readfile)).
```

```js
cy.get('[data-cy="dropzone"]').attachFile('fixtures/myfixture.json', { action: 'drag-n-drop' });
```

#### Overriding the file name

<Badge type="danger">Before</Badge> Dragging and dropping a file with
`cypress-file-upload`

```js
cy.get('[data-cy="dropzone"]').attachFile({
  filePath: 'myfixture.json',
  fileName: 'customFileName.json',
})
```

<Badge type="success">After</Badge> Attaching a fixture from disk with
Cypress 10.0. Cypress follows paths from the root of your test folder

> (as [`cy.readFile()`](/api/commands/readfile)).

Reviewer: ```suggestion
(same as [`cy.readFile()`](/api/commands/readfile)).
```

CONTENT_TBD

### Attaching Files

Attaching files to input elements or dropping them over the page is available
in Cypress 10.0. Read the [`.attachFile()` API docs](/api/commands/attachfile)
for more information on how this works and how to use it.

The
[`cypress-file-upload`](https://github.com/abramenal/cypress-file-upload)

> plugin has been deprecated in favor of `.attachFile()` built into Cypress.

Reviewer 3: We should explicitly state that the cypress-file-upload is incompatible with 10.0, and will no longer work.

Author: Done.

`cypress-file-upload` plugin provided. You can follow the steps below for each argument in order to migrate:

In the first argument:
- `filePath`: Prefix value with `fixtures/`. Rename to `contents`.
- `fileContent`: Use `Buffer.from()` rather than `Cypress.Blob` methods. Rename to `contents`.
- `encoding`: Remove. No longer needed due to improved binary file handling in Cypress 9.0.
- `mimeType`: Remove. No longer needed due to improved binary file handling in Cypress 9.0.

In the second argument:
- `subjectType`: Rename to `action`.
- `allowEmpty`: Remove. `.attachFile()` does not check the length of a file read from disk, only its existence.

### Examples

Below are several examples of migrating various commands from `cypress-file-upload`
to the builtin `.attachFile()`.

#### Read and attach a fixture

<Badge type="danger">Before</Badge> Attaching a fixture from disk with
`cypress-file-upload`

```js
cy.get('[data-cy="file-input"]').attachFile('myfixture.json');
```

<Badge type="success">After</Badge> Attaching a fixture from disk with
Cypress 10.0. Cypress follows paths from your project root
(same as [`cy.readFile()`](/api/commands/readfile)).

```js

> cy.get('[data-cy="file-input"]').attachFile('cypress/fixtures/myfixture.json');

Reviewer 3: Given the note
> `filePath`: Prefix value with `fixtures/`. Rename to `contents`.

Should this example be:
```js
cy.get('[data-cy="file-input"]').attachFile('fixtures/myfixture.json');
```

Author: This section is correct, I've updated the quick guide to match.

CONTENT_TBD

## Migrating `cypress-file-upload` to [`.selectFile()`](/api/commands/selectfile)

Selecting files with input elements or dropping them over the page is available
in Cypress 9.3. Read the [`.selectFile()` API docs](/api/commands/selectfile)
for more information on how this works and how to use it. This guide details how

> to change your test code to migrate from

Reviewer 3: Can we say "migrate from the cypress-file-upload plugin" instead of "migrate from cypress-file-upload" here?

Author: Done.

[`cypress-file-upload`](https://github.com/abramenal/cypress-file-upload) to
`.selectFile()`.

#### Quick guide

The argument signature is different for Cypress' builtin `.selectFile()` command
than the `.attachFile` command the `cypress-file-upload` plugin provided. You
can follow the steps below for each argument in order to migrate:

In the first argument:

- `filePath`: Prefix value with `cypress/fixtures/`. Rename to `contents`.

> - `fileContent`: Use `Buffer.from()` rather than `Cypress.Blob` methods. Rename

Reviewer 3: Do we need to explain to the user where `Buffer` is coming from here? Are they going to be confused?

Author: I've added a link to the nodejs docs for Buffer. Questions like "why can I use Buffer in my cypress code, I thought it executed in the browser" are beyond the scope of a migration guide.

  to `contents`.
- `encoding`: Remove. No longer needed due to improved binary file handling in
  Cypress 9.0.
- `mimeType`: Remove. No longer needed due to improved binary file handling in
  Cypress 9.0.

In the second argument:

- `subjectType`: Rename to `action`. Move from `drag-n-drop` to `drag-drop` or
  from `input` to `select`.
- `allowEmpty`: Remove. `.selectFile()` does not check the length of a file read

>   from disk, only its existence.

Reviewer 3: Could you change this section to something along these lines? I want to be as clear as possible with the language:

When the first argument is a string filepath (or array of string filepaths):

- Prefix the existing filepath(s) with `cypress/fixtures/`.

When the first argument is an object:

- `filePath`: Rename this property to `contents`. Prefix the existing value with `cypress/fixtures/`.
- `fileContent`: Rename this property to `contents`. Use `Buffer.from()` rather than `Cypress.Blob` methods. 
- `encoding`: Remove this property. No longer needed due to improved binary file handling in Cypress 9.0.
- `mimeType`: Remove this property. No longer needed due to improved binary file handling in Cypress 9.0.

The second argument:

- `subjectType`: Rename this property to `action`. Change the existing value from `drag-n-drop` to `drag-drop`, or from `input` to `select`.
- `allowEmpty`: Remove this property. The `.selectFile()` command does not check the length of a file read from disk, only its existence.

Author: Good suggestions, applied.

CONTENT_TBD

## Migrating `cypress-file-upload` to [`.selectFile()`](/api/commands/selectfile)

Selecting files with input elements or dropping them over the page is available
in Cypress 9.3. Read the [`.selectFile()` API docs](/api/commands/selectfile)
for more information on how this works and how to use it. This guide details how
to change your test code to migrate from the
[`cypress-file-upload`](https://github.com/abramenal/cypress-file-upload) plugin
to `.selectFile()`.

> #### Quick guide

Reviewer 3: I feel like all `####` headers in this PR can be changed to `###`, and the `Examples` section header can possibly go away. This would make the right-side nav look more consistent with that of other migration guide sections. Thoughts?

Author: I'm changing quick guide to ###, but I like Examples as a ### because then it shows up in the table of contents, but each individual example doesn't. Since only 2nd and 3rd level headers are rendered:

![image](https://user-images.githubusercontent.com/3003404/148590245-6c98c27a-c50d-4715-aab1-9d798dbe15da.png)

A TOC that looks like

## Migrating from `cypress-file-upload` to [`.selectFile()`](/api/commands/selectfile)
### Quick Guide
### Examples

looks pretty clean while this one is overly verbose for the right-hand column TOC:

## Migrating from `cypress-file-upload` to [`.selectFile()`](/api/commands/selectfile)
### Quick Guide
### Read and attach a fixture
### Using drag-n-drop
### Overriding the file name
### Working with file contents

The argument signature is different for Cypress' builtin `.selectFile()` command
than the `.attachFile` command the `cypress-file-upload` plugin provided. You
can follow the steps below for each argument in order to migrate:

When the first argument is a file path:

> - Prefix the path with `cypress/fixtures`

Reviewer 3: ```suggestion
- Prefix the path with `cypress/fixtures/`.
```

Author: Applied.

CONTENT_TBD

> ## Migrating to Cypress 9.3

Reviewer 3: Instead of having this content in the Migration Guide, which is really geared toward mandatory changes, can you add a section called "Upgrading from the cypress-file-upload plugin" to the `.selectFile()` API docs?

Author: I prefer the solution we used with https://docs.cypress.io/guides/references/migration-guide#Migrating-cy-route-to-cy-intercept, where it's in the migration guide but not attached to a specific release. Moving it into the selectFile docs would clutter up our API docs with a guide that's not of interest to new users or old users once they've gone through it once. It also feels like it would be harder to find in there.

This guide details the changes and how to change your code to make full use of
Cypress 9.3.
[See the full changelog for 9.3](/guides/references/changelog#9-3-0).

> None of the changes detailed here are breaking.

Reviewer 2: I would get rid of this section and just make the main title 'Migrating to selectFile'. This is similar to the section on cy.route to cy.intercept migration which was in a point release.

Author: Ah, that is much nicer. Changing now.

> ### Attaching Files
> 
> Attaching files to input elements or dropping them over the page is available in
> Cypress 9.3. Read the [`.selectFile()` API docs](/api/commands/selectfile) for
> more information on how this works and how to use it.

Reviewer 2: ```suggestion
### Selecting Files to Upload

Selecting files to upload to input elements or dropping them over the page is available in
Cypress 9.3. Read the [`.selectFile()` API docs](/api/commands/selectfile) for
more information on how this works and how to use it.
```

Author: Slightly different tweak since I'm reworking the introduction to match the other style.

The [`cypress-file-upload`](https://github.com/abramenal/cypress-file-upload)
plugin has been deprecated in favor of `.selectFile()` built into Cypress,
though it remains compatible with Cypress 9.3. There's guidance below on how to
migrate from the
[`cypress-file-upload`](https://github.com/abramenal/cypress-file-upload) plugin
to Cypress's built-in selectFile command.

#### Quick guide

> The argument signature is different for Cypress' builtin `.selectFile()` command
> than `.attachFile` the `cypress-file-upload` plugin provided. You can follow the
> steps below for each argument in order to migrate:

Reviewer 2: ```suggestion
The argument signature is different for Cypress' builtin `.selectFile()` command
than the `.attachFile` command the `cypress-file-upload` plugin provided. You can follow the
steps below for each argument in order to migrate:
```

In the first argument:

- `filePath`: Prefix value with `cypress/fixtures/`. Rename to `contents`.
- `fileContent`: Use `Buffer.from()` rather than `Cypress.Blob` methods. Rename
  to `contents`.
- `encoding`: Remove. No longer needed due to improved binary file handling in
  Cypress 9.0.
- `mimeType`: Remove. No longer needed due to improved binary file handling in
  Cypress 9.0.

In the second argument:

- `subjectType`: Rename to `action`. Move from `drag-n-drop` to `drag-drop` or
  from `input` to `select`.
- `allowEmpty`: Remove. `.selectFile()` does not check the length of a file read
  from disk, only its existence.

### Examples

> Below are several examples of migrating various commands from
> `cypress-file-upload` to the builtin `.attachFile()`.

Reviewer 2: ```suggestion
Below are several examples of migrating various commands from
`cypress-file-upload` to the builtin `.selectFile()` command.
```

#### Read and attach a fixture

<Badge type="danger">Before</Badge> Attaching a fixture from disk with
`cypress-file-upload`

```js
cy.get('[data-cy="file-input"]').attachFile('myfixture.json')
```

> <Badge type="success">After</Badge> Attaching a fixture from disk with

Reviewer 2: ```suggestion
<Badge type="success">After</Badge> Selecting a fixture from disk with
```

`.selectFile()`. Cypress follows paths from your project root (same as
[`cy.readFile()`](/api/commands/readfile)).

```js
cy.get('[data-cy="file-input"]').selectFile('cypress/fixtures/myfixture.json')

// Or

cy.fixture('myfixture.json', { encoding: null }).as('myfixture')
cy.get('[data-cy="file-input"]').selectFile('@myfixture')
```

#### Using drag-n-drop

<Badge type="danger">Before</Badge> Dragging and dropping a file with
`cypress-file-upload`

```js
cy.get('[data-cy="dropzone"]').attachFile('myfixture.json', {
  subjectType: 'drag-n-drop',
})
```

> <Badge type="success">After</Badge> Attaching a fixture from disk with

Reviewer 2: ```suggestion
<Badge type="success">After</Badge> Selecting a fixture from disk with
```

`.selectFile()`. Cypress follows paths from the root of your test folder (same
as [`cy.readFile()`](/api/commands/readfile)).

```js
cy.get('[data-cy="dropzone"]').selectFile('fixtures/myfixture.json', {
  action: 'drag-drop',
})
```

#### Overriding the file name

<Badge type="danger">Before</Badge> Dragging and dropping a file with
`cypress-file-upload`

```js
cy.get('[data-cy="dropzone"]').attachFile({
  filePath: 'myfixture.json',
  fileName: 'customFileName.json',
})
```

> <Badge type="success">After</Badge> Attaching a fixture from disk with

Reviewer 2: ```suggestion
<Badge type="success">After</Badge> Selecting a fixture from disk with
```

`.selectFile()`. Cypress follows paths from the root of your test folder (same
as [`cy.readFile()`](/api/commands/readfile)).

```js
cy.get('[data-cy="dropzone"]').selectFile({
  contents: 'fixtures/myfixture.json',
  fileName: 'customFileName.json',
})
```

#### Working with file contents

> <Badge type="danger">Before</Badge> Working with file contents before attaching

Reviewer 2: ```suggestion
<Badge type="danger">Before</Badge> Working with file contents before selecting
```

Author: Picked "using" rather than "selecting" here.

using `cypress-file-upload`

```js
const special = 'file.spss'

cy.fixture(special, 'binary')
  .then(Cypress.Blob.binaryStringToBlob)
  .then((fileContent) => {
    // ...process file contents
    cy.get('[data-cy="file-input"]').attachFile({
      fileContent,
      filePath: special,
      encoding: 'utf-8',
      lastModified: new Date().getTime(),
    })
  })
```

> <Badge type="success">After</Badge> Working with file contents before attaching

Reviewer 2: ```suggestion
<Badge type="success">After</Badge> Working with file contents before selecting
```

Author: Picked "using" rather than "selecting" here.

with `.selectFile()`. The `null` encoding introduced in Cypress 9.0 makes
working with binary data simpler, and is the preferred encoding for use with

> `.attachFile()`.

Reviewer 2: ```suggestion
`.selectFile()`.
```
